scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala

```diff
@@ -373,11 +373,122 @@ object OptimizationRules {
+            Pair(left, _, jf)
+          }
+        case Pair(left, right: CoGrouped[K, v], jf) if forkCoGroup(on, right).isDefined =>
+          forkCoGroup(on, right).map {
+            Pair(left, _, jf)
+          }
+        case Pair(_, _, _) => None // neither side needs a fork
+        case WithDescription(cg, d) => forkCoGroup(on, cg).map(WithDescription(_, d))
+        case WithReducers(cg, r) => forkCoGroup(on, cg).map(WithReducers(_, r))
+        case MapGroup(cg, fn) => forkCoGroup(on, cg).map(MapGroup(_, fn))
+        case FilterKeys(cg, fn) => forkCoGroup(on, cg).map(FilterKeys(_, fn))
+      }
+    }
+
+    /**
+     * The casts in here are safe, but scala loses track of the types in these kinds of
+     * pattern matches.
+     * We can fix it by changing the types on the identity reduces to use EqTypes[V1, V2]
+     * in case class and leaving the V2 parameter.
+     */
+    private def forkReduceStep[A, B, C](on: Dag[TypedPipe], rs: ReduceStep[A, B, C]): Option[ReduceStep[A, B, C]] = rs match {
+      case step@IdentityReduce(_, _, _, _) =>
+        maybeFork(on, step.mapped).map { p => step.copy(mapped = p) }.asInstanceOf[Option[ReduceStep[A, B, C]]]
+      case step@UnsortedIdentityReduce(_, _, _, _) =>
+        maybeFork(on, step.mapped).map { p => step.copy(mapped = p) }.asInstanceOf[Option[ReduceStep[A, B, C]]]
+      case step@IdentityValueSortedReduce(_, _, _, _, _) =>
+        maybeFork(on, step.mapped).map { p => step.copy(mapped = p) }.asInstanceOf[Option[ReduceStep[A, B, C]]]
+      case step@ValueSortedReduce(_, _, _, _, _, _) =>
+        def go(vsr: ValueSortedReduce[A, B, C]): Option[ValueSortedReduce[A, B, C]] =
+          maybeFork(on, step.mapped).map { p =>
+            ValueSortedReduce[A, B, C](vsr.keyOrdering,
+              p, vsr.valueSort, vsr.reduceFn, vsr.reducers, vsr.descriptions)
+          }
+        go(step)
+      case step@IteratorMappedReduce(_, _, _, _, _) =>
+        def go(imr: IteratorMappedReduce[A, B, C]): Option[IteratorMappedReduce[A, B, C]] =
+          maybeFork(on, step.mapped).map { p => imr.copy(mapped = p) }
+        go(step)
+    }
+
+    private def forkHashJoinable[K, V](on: Dag[TypedPipe], hj: HashJoinable[K, V]): Option[HashJoinable[K, V]] =
+      hj match {
+        case step@IdentityReduce(_, _, _, _) =>
+          maybeFork(on, step.mapped).map { p => step.copy(mapped = p) }
+        case step@UnsortedIdentityReduce(_, _, _, _) =>
+          maybeFork(on, step.mapped).map { p => step.copy(mapped = p) }
+        case step@IteratorMappedReduce(_, _, _, _, _) =>
+          maybeFork(on, step.mapped).map { p => step.copy(mapped = p) }
+      }
+
     def apply[T](on: Dag[TypedPipe]) = {
-      case Fork(_) | ForceToDisk(_) => None // these are already forking
-      case SourcePipe(_) => None // don't need to worry about sources
-      case IterablePipe(_) => None // don't need to worry about sources
-      case nonFork if on.fanOut(nonFork) > 1 => Some(Fork(nonFork))
```

Comment:
this was unsafe. If `Fork(nonFork)` already exists, then replacing `nonFork` with `Fork(nonFork)` creates an infinite loop.